src/index.js

```diff
@@ -80,23 +92,64 @@ var tracker = function (w) {
     }, t)
   }
 
+  function saveLastActivityTimestamp() {
+    storage.setItem(STORAGE_LAST_TIMESTAMP, JSON.stringify(getUTCNow()))
+  }
+
   function addEvent(eventName, eventProps) {
+    var timestamp = getUTCNow()
     var ev = eventProps || {}
     ev['event'] = eventName || ''
-    ev['timestamp'] = dateFormatted()
-    ev['session_start'] = session
-    ev['uuid'] = userCookie
-
+    ev['timestamp'] = utcToFormattedDate(timestamp)
+    ev['session_start'] = utcToFormattedDate(sessionStart)
+    ev['uuid'] = uuid
     events.push(ev)
+
+    saveLastActivityTimestamp()
   }
 
   function die() {
     storage.setItem(STORAGE_ITEM, JSON.stringify(events))
     uploadEvents()
   }
 
+  function refreshSessionStart() {
+    var now = getUTCNow()
+    var lastActivityTimestamp = storage.getItem(STORAGE_LAST_TIMESTAMP)
+    if (lastActivityTimestamp !== null) {
+      var elapsed = now - parseInt(lastActivityTimestamp, 10)
+      if (elapsed > REFRESH_SESSION) {
+        sessionStart = now
+        setCookie(COOKIE_NAME, formatCookieValue(uuid, sessionStart))
+        saveLastActivityTimestamp()
+      }
+    }
+  }
+
+  var hidden, visibilityChange
+  if (typeof doc.hidden !== 'undefined') {
+    // Opera 12.10 and Firefox 18 and later support
+    hidden = 'hidden'
+    visibilityChange = 'visibilitychange'
+  } else if (typeof doc.msHidden !== 'undefined') {
+    hidden = 'msHidden'
+    visibilityChange = 'msvisibilitychange'
+  } else if (typeof doc.webkitHidden !== 'undefined') {
+    hidden = 'webkitHidden'
+    visibilityChange = 'webkitvisibilitychange'
+  }
+
+  function handleVisibilityChange() {
+    if (!doc[hidden]) {
+      refreshSessionStart()
+    }
+  }
+
   w.addEventListener('beforeunload', die)
   w.addEventListener('unload', die, false)
+  doc.addEventListener(visibilityChange, handleVisibilityChange, false)
+
+  refreshSessionStart()
 
   // Overwritte tinybird function
   var queue = w.tinybird || []
```

Comment:
Not from this PR but...

I don't understand why this is like it is. Why window.tinybird is an array of events. Where does that happen? I'm confused about this code. @xavijam

Reply:
You can add events in the `tinybird` array before the script is loaded.